crates/bevy_render/src/mesh/mesh/mod.rs

```diff
@@ -544,12 +544,37 @@ impl Mesh {
     }
 
     /// Calculates the [`Mesh::ATTRIBUTE_NORMAL`] of a mesh.
+    /// If the mesh is indexed, this defaults to smooth normals. Otherwise, it defaults to flat
+    /// normals.
     ///
     /// # Panics
     /// Panics if [`Mesh::ATTRIBUTE_POSITION`] is not of type `float3`.
     /// Panics if the mesh has any other topology than [`PrimitiveTopology::TriangleList`].
     ///
-    /// FIXME: The should handle more cases since this is called as a part of gltf
+    /// FIXME: This should handle more cases since this is called as a part of gltf
+    /// mesh loading where we can't really blame users for loading meshes that might
+    /// not conform to the limitations here!
+    pub fn compute_normals(&mut self) {
+        assert!(
+            matches!(self.primitive_topology, PrimitiveTopology::TriangleList),
+            "`compute_normals` can only work on `TriangleList`s"
+        );
+        if self.indices().is_none() {
+            self.compute_flat_normals();
+        } else {
+            self.compute_smooth_normals();
+        }
+    }
+
+    /// Calculates the [`Mesh::ATTRIBUTE_NORMAL`] of a mesh.
+    /// If the mesh is not indexed, there must not be any shared vertices. Consider calling
```

Comment:
Isn't that the other way around? As in, duplicate_vertices should be called on indexed meshes? Non indexed meshes should just work as is because it means they already don't share vertices?

Reply:
I was trying to address @ManevilleF's comment:
> We could add a doc comment explaining that a certain layout of vertices are expected for non indexed geometry, similar to the output of duplicate_vertices

but reading this again, it sounds like this functionality is encapsulated by the fact that this method branches based on whether the mesh is indexed. Calling `duplicate_vertices` on an unindexed mesh just to call `compute_flat_normals` seems redundant, and it would also increase the vertex count. What do you think about just removing this sentence from the doc?

Reply:
I meant the indexed path. The unindexed path doesn't care about duplicating vertices because it's already not sharing any vertices. The indexed path is the one that needs to call duplicate_vertices. That why I wanted to keep the original function without trying to support indexed and non-indexed. The original only supported non-indexed meshes because that's the easiest to implement. Indexed meshes need duplicate vertices to compute flat normals or at least, that's how it worked in 0.13 and anything before that.

That's what the assert does at the top of the original function,https://github.com/bevyengine/bevy/blob/latest/crates/bevy_render/src/mesh/mesh/mod.rs#L535 it panics if there are indices and suggests to use duplicate_vertices.

We really need some test for these things 😅 (to be clear, you don't need to add tests, but that would have made things like this a bit clearer) .

Reply:
> The indexed path is the one that needs to call duplicate_vertices

Is there some additional functionality outside the methods that are getting updated in this PR that requires indexed meshes to have duplicated vertices? My understanding is that the previous implementation required users to call `duplicate_vertices` specifically so that flat normals could be computed.

I'm still failing to understand how calling `duplicate_vertices` on an indexed mesh would break `compute_flat_normals` 😖. Doesn't the addition of [this bit of code](https://github.com/bevyengine/bevy/pull/13333/files#diff-dd6c1ace6c4248856ec7e9f21ddb1791ececdf1484e61a9a52dcf62aeab96a68R592-R595), which branches depending on whether the indices are defined, result in us referencing the correct vertices in the [subsequent `chunk_exact` call](https://github.com/bevyengine/bevy/pull/13333/files#diff-dd6c1ace6c4248856ec7e9f21ddb1791ececdf1484e61a9a52dcf62aeab96a68R598-R603)?

Tangentially, I'm happy to create a new issue to add some unit testing to this and pick that up as a follow-on (separate from this PR).

Reply:
> My understanding is that the previous implementation required users to call duplicate_vertices specifically so that flat normals could be computed.

Exactly, but now in this PR compute_flat_normals just assumes that users have done that. So it's possible that users calls compute_flat_normals on indexed meshes without duplicating vertices and it will not behave as expected and without warnings.

> I'm still failing to understand how calling duplicate_vertices on an indexed mesh would break compute_flat_normals

It wouldn't. The issue is _not_ calling it on an indexed mesh and then trying to compute flat normals. Currently, in this PR an indexed mesh without duplicated vertices is accepted with no warnings.

Here's how computing flat normals on an icosphere with indices looks like with this PR

no duplicated vertices:
![image](https://github.com/bevyengine/bevy/assets/8348954/2dcfb495-cb2e-4c3c-8311-d5eedfa0f5db)

duplicated vertices:
![image](https://github.com/bevyengine/bevy/assets/8348954/bcfaa4ad-c030-4a1e-9887-3ce0c8c8c448)

The no duplicated vertices case should not be valid, but this PR currently ignores this case without any warnings _and_ the comment about calling `duplicate_vertices` is inverted from what should be done. 

There's no easy way to detect if a mesh has duplicated vertices though, so at the very least the doc comment should be corrected to say it's only needed when the mesh is indexed. 

Also, if a mesh _does_ call `duplicate_vertices` then we can just always use the positions to compute the flat normals which means you don't need to handle the case of indexed meshes. So the previous behaviour of only computing it on positions and panic on indexed meshes still seems more correct to me.

Reply:
Another note, `duplicate_vertices` consumes the indices. So essentially, you'd never end up in a case where you need indices to compute flat normals.

Either you have a non-indexed mesh and you're fine, or you have an indexed mesh, in which case you need to call duplicate_vertices always because you can't correctly compute flat normals without it.

Reply:
I guess the one correct situation where calling compute_flat_normals on an indexed mesh _without_ calling duplicate_vertices is when someone created an indexed mesh where each index is a single vertex, but at that point using indices is pointless so it's a very niche edge case. In most cases indexed meshes aren't like that.

Reply:
Ah ok, you've helped me dust off some cobwebs from my understanding of graphics programming 😅. I agree that it makes more sense to revert `compute_flat_normals` to the previous behavior where we panic on indexed meshes.

Reply:
Happy to help, sorry, I think some of my original comments were a bit misleading and increased the confusion. Anyway, I think you can resolve this now.